pyscf/lib/linalg_helper.py

```diff
@@ -1384,72 +1386,37 @@ def krylov(aop, b, x0=None, tol=1e-10, max_cycle=30, dot=numpy.dot,
             for j in range(nd):
                 h[i,j] = dot(xi.conj(), ax[j])
         xi = None
-
     # Add the contribution of I in (1+a)
     for i in range(nd):
         h[i,i] += innerprod[i]
 
     g = numpy.zeros((nd,nroots), dtype=x1.dtype)
-    if b.ndim == 1:
-        g[0] = innerprod[0]
-    else:
-        # Restore the first nroots vectors, which are array b or b-(1+a)x0
-        for i in range(min(nd, nroots)):
-            xsi = numpy.asarray(xs[i])
-            for j in range(nroots):
-                g[i,j] = dot(xsi.conj(), b[j])
+    for i in range(nd):
+        g[i] = dot(b, numpy.asarray(xs[i]).conj())
 
     c = numpy.linalg.solve(h, g)
     x = _gen_x0(c, xs)
     if b.ndim == 1:
         x = x[0]
 
+    # Restore the first nroots vectors, which are array b or b-(1+a)x0
     if x0 is not None:
         x += x0
     return x
 
 
-def solve(aop, b, precond, tol=1e-12, max_cycle=30, dot=numpy.dot,
+def solve(aop, b, precond=None, tol=1e-12, max_cycle=60, dot=numpy.dot,
           lindep=DSOLVE_LINDEP, verbose=0, tol_residual=None):
-    '''Davidson iteration to solve linear equation.
+    '''Solve linear equation using the scipy.sparse module
```

Comment:
maybe specify that GMRES is used, or we could rename the function as something like `gmres_wrapper`